### PR TITLE
Plumb let bindings through scope checking

### PR DESCRIPTION
**Pull request by harlanhaskins**
### What's in this pull request?

This patch implements scope checking for `let` bindings (module type checker bugs), and adds appropriate tests.

Along the way, we fixed some issues with our representation of arrows
and cleaned up a few sections.

### CODE REVIEW

File: Sources/Mantle/Elaborate.swift [hunk 1]
```diff
@@ -245,25 +251,6 @@ extension TypeChecker where PhaseState == ElaboratePhaseState {
     return t
   }
 
-  // Takes a Pi-type and replaces all it's elements with metavariables.
-  private func fillPiWithMetas(_ ty: Type<TT>) -> [Term<TT>] {
```
Comment (CodaFi) [inline Sources/Mantle/Elaborate.swift]: Could you revert this change?  I thought I'd need it at the time but now I realize I don't.

File: Sources/Mantle/Signature.swift [hunk 2]
```diff
@@ -135,6 +135,14 @@ extension Signature {
                                             inside: newDef))
   }
 
+  func addLetBinding(_ name: QualifiedName, type: Type<TT>,
+                     _ tel: Telescope<TT>) {
+    let definition = Definition.constant(type, .function(.open))
```
Comment (CodaFi) [inline Sources/Mantle/Signature.swift]: Just a thought: Maybe we should add let bindings as a new kind of constant.  Not sure it's worth the effort.

File: Sources/Moho/Reparse.swift [hunk 3]
```diff
@@ -508,6 +498,15 @@ extension NameBinding {
   ///   (_-_ (_+_ n n) n)
   /// ```
   func reparseExpr(_ syntax: ExprSyntax) -> ExprSyntax {
+    switch syntax {
+    case let syntax as LetExprSyntax:
+      return syntax
```
Comment (CodaFi) [inline Sources/Moho/Reparse.swift]: Notation can occur in the RHS of a let expression.  This should be

```swift
return syntax.withOutputExpr(self. reparseExpr(syntax.outputExpr))
```

Comment (CodaFi) [reply Sources/Moho/Reparse.swift]: Also needs a test that uses notation in that position to validate

File: Sources/Moho/Syntax.swift [hunk 4]
```diff
@@ -97,6 +97,12 @@ public enum Decl {
   /// defined in the current module.  By now, any conflicts this might cause
   /// have been resolved by Scope Check.
   case openImport(QualifiedName)
+  /// A function declaration without a signature in a `let` binding.
```
Comment (CodaFi) [inline Sources/Moho/Syntax.swift:100]: Not just functions, any (local) declaration!

**Comment by CodaFi** [comment]
I think this patch does too much in its current form.  I'd like to see the improvements to ancillary components split from the TC changes so we can isolate the XFailure

**Review by CodaFi — APPROVED** [review]

**Comment by harlanhaskins** [comment]
Tests passed on Linux. ⛴
